src/eradiate_plugins/bsdfs/rtls.cpp

```diff
@@ -0,0 +1,322 @@
+    }
+
+    std::pair<BSDFSample3f, Spectrum> sample(const BSDFContext &ctx,
+                                             const SurfaceInteraction3f &si,
+                                             Float /* position_sample */,
+                                             const Point2f &direction_sample,
+                                             Mask active) const override {
+        MI_MASKED_FUNCTION(ProfilerPhase::BSDFSample, active);
+
+        Float cos_theta_i = Frame3f::cos_theta(si.wi);
+        BSDFSample3f bs   = dr::zeros<BSDFSample3f>();
+
+        active &= cos_theta_i > 0.f;
+        if (unlikely(dr::none_or<false>(active) ||
+                     !ctx.is_enabled(BSDFFlags::GlossyReflection)))
+            return { bs, 0.f };
+
+        bs.wo           = warp::square_to_cosine_hemisphere(direction_sample);
+        bs.pdf          = warp::square_to_cosine_hemisphere_pdf(bs.wo);
+        bs.eta          = 1.f;
+        bs.sampled_type = +BSDFFlags::GlossyReflection;
+        bs.sampled_component = 0;
+
+        UnpolarizedSpectrum value =
+            eval_rtls(si, bs.wo, active) * Frame3f::cos_theta(bs.wo) / bs.pdf;
+
+        return { bs, depolarizer<Spectrum>(value) & (active && bs.pdf > 0.f) };
+    }
+
+    inline const UnpolarizedSpectrum eval_K_iso() const { return 1.0f; }
+
+    inline const UnpolarizedSpectrum eval_K_vol(const Float &cos_theta_i,
+                                                const Float &cos_theta_o,
+                                                const Float &cos_psi,
+                                                const Float &sin_psi,
+                                                const Float &psi) const {
+        return ((dr::Pi<Float> / 2.f - psi) * cos_psi + sin_psi) /
+                   (cos_theta_i + cos_theta_o) -
+               (dr::Pi<Float> / 4.f);
+    }
+
+    inline const UnpolarizedSpectrum eval_D(const Float &tan_theta_i,
+                                            const Float &tan_theta_o,
+                                            const Float &cos_d_phi) const {
+        return dr::sqrt(dr::sqr(tan_theta_i) + dr::sqr(tan_theta_o) -
+                        2.f * tan_theta_i * tan_theta_o * cos_d_phi);
+    }
+
+    inline const UnpolarizedSpectrum eval_O(const Float &tan_theta_i,
+                                            const Float &tan_theta_o,
+                                            const Float &sec_theta_sum,
+                                            const Float &cos_d_phi,
+                                            const Float &sin_d_phi) const {
+
+        const UnpolarizedSpectrum D =
+            eval_D(tan_theta_i, tan_theta_o, cos_d_phi);
+        Log(Trace, "D: %s", D);
+
+        const Float tan_sin_prod = tan_theta_i * tan_theta_o * sin_d_phi;
+        UnpolarizedSpectrum cos_t =
+            (m_h / m_b) * dr::sqrt(dr::sqr(D) + dr::sqr(tan_sin_prod)) /
+            sec_theta_sum;
+
+        // Clip cos(t) values outside of [-1; 1]
```

Comment:
Maybe here you can comment on when you expect this variable to take such values.